Fix for outstanding riak_kv/679 dataloss cases

russelldb opened the pull request:
These commits fix the remaining kv679 dataloss edge cases. They include a simple quickcheck model/simulator. The riak-tests for these changes can be found here https://github.com/basho/riak_test/pull/1296.

The fix is a slight change to the per-actor-epoch code that shipped in 2.1, whereas before we only updated the epoch on coordinating writes, now we update it on non-coordinating writes where we detect some local amnesia. The fix is to add a no-op/no-dot entry to the version vector with a new actor-epoch and a counter of one. This ensures that the next write coordinated by the vnode uses the new epoch, and is not instead a repeated event.

russelldb committed:
A new key epoch is only needed when coordinating

new_key_epoch was called for all puts where the local get returned a
notfound (no_old_object in the parlance of the recent refactor.) A new
epoch is only needed in certain situtions. This change means a new epoch
is only created when needed on no_old_object puts.

This is an intermediate step toward fixing the outstanding kv679 case
where a vnode has a stale/old/bad local vclock on disk. See the test
https://github.com/russelldb/riak_test/blob/rdb/gh-kv679-fb-byz/tests/kv679_dataloss_fb.erl
for more details.

This commit also simplifies the maybe_new_key_epoch function.

russelldb committed:
Slight refactor for readability

And to dodge this weird erlang bug
http://erlang.org/pipermail/erlang-questions/2017-March/091844.html

russelldb committed:
Add basic failing eqc for kv679 bugs

In this test all writes are coordinated by the head of the preflist,
pfhead, and replicated to at least one other node chosen from 2 further
primaries and 3 fallbacks. Every write is there for w=2 and acked on
pfhead one other node. Every now and again one node may replicate to
another. Every now and again pfhead may "forget" it's local values (this
is like a disk error or partially successful delete.) In this commit the
test catches the lost acked write bug pretty quickly. Next up is to add
the vnode counter and see if the test can catch the case as yet unfixed.

Also added a convenience function to riak_object.

russelldb committed:
Add basic coordinating epoch counter

This commit adds the epoch counter to the coordinator. It's a WIP commit
that fixes the counter example from the previous commit (which showed
dataloss.)

russelldb committed:
WIP: counter example that matches outstanding bug

This counter example shows the same bug as the riak_test
https://github.com/basho/riak_test/blob/rdb/gh-kv679-fb-byz/tests/kv679_dataloss_fb.erl
Yay!

russelldb committed:
Add epochs to non-coordinating writes

This commit adds a no-op/no-dot VV increment with a new epoch actor to
an object if it is needed. If the incoming object contains an entry
for the vnode greater than the local one then that indicates that the
local vnode has suffered amnesia and needs to ensure it uses a new
actor id next time it coordinates a write on this key. The simplest
way to do that is add a VV entry of Actor:Epoch+1 with a counter of 1.
Note the new function on riak_object as this increment must _NEVER_
create a dot.

russelldb committed:
Add counter example for different coordinator

Contrary to previous commit, there is a variation on the bug that
requires new actor epoch even if the local object is found.

russelldb committed:
Fix case where a vnode stores own stale write

An amnesiac vnode may recieve a replica/handoff put with it's own vnode
ID in, this is a strong hint that something is rotten in Denmark, add a
new vnode epoch to the clock as a non-dot/no-op entry

russelldb committed:
WIP: maybe fix to non-coord kv679

I think this all needs a refactor.

russelldb committed:
Add a counter example

russelldb committed:
WIP: fix for dataloss bug

russelldb committed:
Clean up some comments, add some specs

is_local_amnesia seems a better function name than is_greater

russelldb committed:
Add some comments

JeetKunDoug commented:
@russelldb if you can start by addressing the dialyzer warnings while Nick and I are reviewing I'd appreciate it. On first pass, makes sense to me and generally looks good, but I want to think a bit more about it. There are a few stylistic things we're trying to avoid in here, but nothing we can't clean up post-merge.

russelldb commented:
Happy to address the stylistic concerns if you air them.

Did I introduce new dialyzer warnings?

russelldb commented:
Ah, I see it, thumbs is pretty useful/helpful.

JeetKunDoug commented:
Yea - we ❤️ Thumbs these days... only real issues are where we may have set some of the settings in a way that we later figured out we didn't like...

I was just running dialyzer on `develop` to make sure there isn't anything outstanding there before replying to you but looking at the errors I'm fairly certain they are from this commit (and can now confirm that is true).

JeetKunDoug commented on src/riak_kv_vnode.erl:
```diff
@@ -1681,29 +1688,23 @@ select_newest_content(Mult) ->
-            {oldobj, CurObj};
-        false ->
-            {newobj, ResObj}
+    if NewEpoch ->
```
please replace this with a `case NewEpoch` - trying to kill `if` statements with 🔥 these days, and would like to get rid of the ?ELSE macro as well (but understand that the other usage isn't currently part of this PR, so it's OK to leave it for now).

nickelization commented on src/riak_kv_vnode.erl:
```diff
@@ -1512,7 +1512,7 @@ prepare_put_new_object(#state{idx =Idx} = State,
                          []
                  end,
-    {EpochId, State2} = new_key_epoch(State),
+    {EpochId, State2} = maybe_new_key_epoch(Coord, State, undefined, RObj),
```
Instead of using `undefined` here, could we use a more distinctive atom like `no_object` or something along those lines? Would make the code a bit clearer, and also help guard against any future bugs where an unrelated use of `undefined` might leak into this code. Admittedly this is somewhat of a personal style preference though, so I won't argue if you'd strongly prefer leave it as-is. Looking through the code, we'd have to update `maybe_new_key_epoch` and also `highest_actor`, but I think it should be a safe and easy change.

(Commenting on the first commit where this appears, but it does look like this code later got moved into `maybe_update_vclock`.)

nickelization commented on src/riak_kv_vnode.erl:
```diff
@@ -2598,47 +2598,40 @@ non_neg_env(App, EnvVar, Default) when is_integer(Default),
+-spec maybe_new_key_epoch(Coord::boolean(), State::#state{},
+                          LocalObject::riak_object:riak_object(),
+                          IncomingObject::riak_object:riak_object()) ->
+                                 {ActorId:: binary(), #state{}}.
```
Thanks for adding the variable names to the type specs - definitely improves readability in many places.

In the case of `ActorId` I see we specify `ActorId :: binary()` in a couple places. Maybe worth defining a local `actor_id()` type in this module to avoid redundancy?

russelldb replied:
Ah, I found an existing type `vnodeid()` but I'm not sure if it muddies the water since in some places that means the one generated by the vnode_status_mgr, and some cases it means the actor_id generated by smushing together the vnodeid and epoch. Maybe we need both, using the bitstring <<_:Length>> type spec?

nickelization commented on src/riak_kv_vnode.erl:
```diff
@@ -2656,17 +2649,19 @@ key_epoch_actor(ActorBin, Cntr) ->
+%% `ActorBase' that has acted on this key. `KeyEpoch' is the highest
+%% epoch for the `ActorBase'. `Counter' is the greatest event seen by
+%% the `VnodeId'.
 -spec highest_actor(binary(), riak_object:riak_object()) ->
```
Since we are now allowed to pass `undefined` into this function, we should add it to the type spec.

nickelization commented:
@russelldb what happened? I was about to continue my review, and was excited to get this stuff merged.

russelldb commented:
I have no idea, I didn't do anything!

nickelization commented:
Wow, weird! Well, glad it was nothing then, haha.

russelldb commented:
Figured it out, I had a script running that did a push remote --mirror but the local repo did not have the branch so it deleted it, sorry for the confusion

nickelization commented:
Ahh gotcha, no worries, makes sense.

russelldb commented:
in the past all my PRs where from a basho branch, not a fork :)

nickelization commented on src/riak_kv_vnode.erl:
```diff
@@ -2598,75 +2629,95 @@ non_neg_env(App, EnvVar, Default) when is_integer(Default),
-    {State#state.vnodeid, State};
-maybe_new_key_epoch(true, State=#state{counter=#counter_state{use=false}, vnodeid=VId}, _, _) ->
+-spec maybe_new_key_epoch(Coord::boolean(), State::#state{},
+                          LocalObject::riak_object:riak_object(),
```
This type spec should include `undefined` as a possible value for `LocalObject` (or whatever other placeholder we end up using to mean "no local object" if we end up changing the code to use a different atom).

nickelization commented on test/kv679_eqc.erl:
```diff
@@ -0,0 +1,493 @@
+eqc_test_() ->
+    {setup,
+     fun() ->
+             meck:new(riak_core_bucket),
```
We could just call `setup()` and `teardown()` here instead of repeating the bodies of the functions.

russelldb replied:
Will do, I added `setup` and `teardown` after the `eqc_test` function, didn't refactor.

nickelization commented on test/kv679_eqc.erl:
```diff
@@ -0,0 +1,493 @@
+    %% end.
+
+%% like vclock:timestamp() but monotinically increasing
+timestamp() ->
```
This function will not return monotonically increasing values (or even necessarily unique ones) since we're adding the microseconds field directly to the seconds field without any scaling. I'm not sure if this is critical to the correctness of this test though, will have to think about it a bit more.

russelldb replied:
It's my mistake, I was meant to do something smarter but never finished. Will address. So far the test has worked, but that might be good luck.

nickelization commented on test/kv679_eqc.erl:
```diff
@@ -0,0 +1,493 @@
+set_up_replicas() ->
+    [ ets:insert(?MODULE, #replica{id=Id}) || Id <- ?REPLICAS ++ [?P1] ].
+
+%% two dicts come in, a true or a dict comes out, you can't explain
```
😆

nickelization commented on test/kv679_eqc.erl:
```diff
@@ -0,0 +1,493 @@
+-record(state, {
+          pfhead=?P1,
+          %% Actor Ids for replicas in the system
+          replicas= ?REPLICAS,
```
Seems like this is never modified, right? Would be slightly simpler to just use the `?REPLICAS` constant directly in the code and kill this state field. Though if there are future plans to make this field mutable then it might be worth keeping it around.

EDIT - Same with `pfhead` too.

russelldb replied:
yeah, these are left over from earlier iterations.

nickelization commented:
I've picked over all this code with a fine-tooth comb, and other than the few minor things I noted in my comments, it looks great. Solid stuff! @JeetKunDoug is setting up a Giddyup run with this code to check for any regressions, but assuming that all passes, this gets a +1 from me.

russelldb committed:
Address dialyzer warnings

There is one still outstanding, but it is baffling, and, I suspect
predates these changes.

russelldb commented:
Many thanks for the review. I'm working on it. One dialyzer warning remains, I have had to post to the erlang ML for help.

russelldb commented:
Re-running the eqc I've found some new counter examples. Hold off on this until I figure them out, please.

russelldb committed:
Address review comments, and fix generator

The generator for put has to ensure that W=2 and that the preflist is
distinct nodes.

Also, for some reason the epoch update on non-coordinating writes was
commented out, so I re-instated that, as it should be there.

nickelization commented:
@russelldb Was chatting with @JeetKunDoug and he mentioned a test failure during the Giddyup run for `partition_repair`. I checked it out locally, and looks like it fails consistently (and passes on develop). Haven't done any real investigation into the failure, and I'm about to sign off for the day, but I'll try and take a look tomorrow morning if you don't get a chance to check it out yourself before then.

russelldb commented:
@nickelization looking at it. So far a bit baffled. All the `norfound` objects are from the same partition, partition `0`. Hopefully I'll have a better idea by the time you're online

nickelization commented:
@russelldb been a busy day and I haven't had much chance to look into this any further yet (I probably should've mentioned earlier, but today is my last day working for Basho, so I've had some loose ends to tie up). The one potential lead I have for you via Doug is that he noticed those new "amnesia" log messages showing up during repairs. I guess that might not be unexpected, due to the nature of the test...but again, I haven't had a chance to look into it any further yet, and might not get to it before my time here is up :(

russelldb committed:
Fix dialyzer warnings

With thanks to Dough Rohrer. `new_key_epoch/1` is only ever called if
`#counter_state.use == true`.

Also address stylistic complaints about use of `if ?ELSE`

russelldb commented:
@nickelization wow. Well thanks for all the time you put into this PR up to now. All the best in whatever you're doing next. Stay in touch, please! Hope we can work together again in the future.

russelldb commented:
@JeetKunDoug Finally figured it out. https://github.com/basho/riak_test/blob/develop/tests/partition_repair.erl#L223 is the problem

The data in the vnode partition no longer matches the data in the stash since it now has the new epoch actor entry added to the vclock. I guess I need to makes this `verify` function smarter, in that each key in the partition after repair must have the same value, and _at least_ the vclock of the stash. Sound about right?

JeetKunDoug commented:
@russelldb yea - that looks about right to me.. a more intelligent comparison should do it. LMK if you need anything else, or when you've got changes done, and I'll take another look.

russelldb commented:
@JeetKunDoug I pushed a fix to the riak-test at https://github.com/basho/riak_test/pull/1296

JeetKunDoug commented:
@russelldb thanks - it seemed like yo had solutions for the dialyzer errors - any clues on the eunit failure in the Thumbs output? Looks potentially unrelated but curious if you have taken a look yet? If not I can try to see if it's reproducible locally on develop or not and we can go from there.

nickelization commented:
@russelldb you're welcome, and thanks! Same to you, on all counts :)

russelldb commented:
@JeetKunDoug I used your solutions for the dialyzer errors! Will look at the eunit

russelldb committed:
Fix failing eqc to use new put_merge API

riak_kv_vnode:put_merge is exported for testing, and the kv679 changes
change the function signature from `vnodeid` to a two-tuple
`{IsNewEpoch::boolean(), VnodeId::binary()}` this commit reflects that
change in the test code.

russelldb commented:
@JeetKunDoug Fixed the failing eunit. `fsm_qc_vnode` uses `riak_kv_vnode:put_merge`. That 

> %% TODO: The riak_kv_vnode code should be refactored to expose this function
> %%       so we are close to testing the real thing.

is looking more and more pertinent as time goes on :D

russelldb committed:
Final Dialyzer fixen

Thanks, again, to @JeetKunDoug

JeetKunDoug commented:
Now that all the unit test/dialyzer issues are taken care of, going to kick off another integration test run... should have results in the morning.

JeetKunDoug commented:
+1 - integration tests run and triaged. Looks good. Thanks @russelldb for the great work, as always.

angrycub commented:
+1

JeetKunDoug commented:
@thumbot merge

russelldb commented:
Wooo! Thanks @JeetKunDoug and @nickelization for the review.

davidx commented:
@thumbot settings
